cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,442 @@
+
+![SPIFFE and SPIRE Architecture with Cilium Datapath](./images/SPIFFE-SPIRE-arch-with-datapath.png)
+
+Steps following an authentication map lookup failure:
+1. The packet is dropped. The BPF datapath emits a drop event to userspace notifying that the packet between these peers has been dropped due to the reason that "authentication is required" for this connection.
+1. Cilium agent intercepts the event and initiates a request for certificate material that is necessary in order to establish a TLS session with the cilium-agent on the node where the peer application resides.
+1. Cilium agent initiates the mTLS handshake with the peer node
+1. Cilium agent populates the BPF auth map to indicate that these peers have authenticated, and as such subsequent packets between these peers can perform the lookup in this map and pass the policy/auth check and hence continue to the destination.
+
+The above outline describes the process at the source node. A similar process needs to occur at the destination node, to allow the destination node to authenticate the source pod. A naive approach that should work is to execute the exact same logic at the destination node when packets are processed at ingress towards the destination pod. An improved implementation would initiate this process at the destination node when the source node initiates step (3) towards the destination node.
+
+### Debuggability
+* In-Pod CLI, extend cilium bugtool to gather this data so it gets into sysdumps:
+  * cilium status - overview of authentication status
+    * N/M pods configured for authentication
+    * N authentication sessions with peers
+  * cilium auth list - List all authentication sessions in cilium-agent
+    * Local node
+    * Peer node
+    * Local pod (selector?)
+    * Peer pod (selector?)
+    * Last authentication timestamp
+    * ETA next authentication timestamp
+    * Inbound / Outbound session establishment (i.e. triggered locally or by remote peer)
+    * Status (authenticating / authenticated / failed + err )
+  * cilium bpf auth list - Dump bpf map state for authentication (if applicable)
+* hubble events - Add "authenticated" timestamp to events so that hubble CLI, UI, etc can visualize authentication status of connections between in-cluster peers
+  * Perhaps, "last authentication time" + "next authentication time"?
+  * Was authentication required by the local peer or the remote peer for this traffic?
+  * Discussion needed: SIG-Hubble
+* Metrics
+  * Total authentication sessions in the cilium-agent (inbound vs. outbound)
+  * SPIFFE ID registration time (during CNI ADD)
+  * TLS Session authentication time (during connection establishment) 
+  * Your metrics suggestions here
```

Comment:
* we have a mechanism to check if a received SVID is new or an update, Worth exposing this?
* the number of failed mTLS handshakes and their reason? (missing peer cert etc?)

Reply:
Great, added!